CMake/curl-config.cmake.in

```diff
@@ -49,3 +48,30 @@ endif()
 # For compatibility with CMake's FindCURL.cmake
 set(CURL_LIBRARIES @PROJECT_NAME@::libcurl)
 set_and_check(CURL_INCLUDE_DIRS "@PACKAGE_CMAKE_INSTALL_INCLUDEDIR@")
+
+set(CURL_VERSION_STRING "@CURLVERSION@")  # for compatibility
+
+set(CURL_SUPPORTED_PROTOCOLS "@CURL_SUPPORTED_PROTOCOLS_LIST@")
+set(CURL_SUPPORTED_FEATURES "@CURL_SUPPORTED_FEATURES_LIST@")
+
+foreach(_item IN LISTS CURL_SUPPORTED_PROTOCOLS CURL_SUPPORTED_FEATURES)
+  set(CURL_SUPPORTS_${_item} TRUE)
+endforeach()
+
+set(_anymissing FALSE)
+foreach(_item IN LISTS CURL_FIND_COMPONENTS)
+  if(CURL_SUPPORTS_${_item})
+    set(CURL_${_item}_FOUND TRUE)  # for compatibility
+  else()
+    if(CURL_FIND_REQUIRED_${_item})
+      set(CURL_FOUND FALSE)
```

Comment:
Thanks for your suggestions and code! Happy New Year to you too :)

I adatped your changes with a few small changes. Then when testing
I noticed that CMake is already showing the list of passed/missed
components in its built-in "found" message. So I dropped two of them.
When combined with `REQUIRED` it would only show the first missed
components, so kept that one to show all missing ones.

Ending up with:
```cmake
# For compatibility with CMake's FindCURL.cmake
set(CURL_VERSION_STRING "@CURLVERSION@")
set(CURL_LIBRARIES @PROJECT_NAME@::libcurl)
set_and_check(CURL_INCLUDE_DIRS "@PACKAGE_CMAKE_INSTALL_INCLUDEDIR@")

set(CURL_SUPPORTED_PROTOCOLS "@CURL_SUPPORTED_PROTOCOLS_LIST@")
set(CURL_SUPPORTED_FEATURES "@CURL_SUPPORTED_FEATURES_LIST@")

foreach(_item IN LISTS CURL_SUPPORTED_PROTOCOLS CURL_SUPPORTED_FEATURES)
  set(CURL_SUPPORTS_${_item} TRUE)
endforeach()

set(_missing_req "")
foreach(_item IN LISTS CURL_FIND_COMPONENTS)
  if(CURL_SUPPORTS_${_item})
    set(CURL_${_item}_FOUND TRUE)
  elseif(CURL_FIND_REQUIRED_${_item})
    list(APPEND _missing_req ${_item})
  endif()
endforeach()

if(_missing_req)
  string(REPLACE ";" " " _missing_req "${_missing_req}")
  if(CURL_FIND_REQUIRED)
    message(FATAL_ERROR "CURL: missing required components: ${_missing_req}")
  endif()
  unset(_missing_req)
endif()

check_required_components("CURL")
```

used this script for testing:
```shell
#!/bin/sh

mv $PWD/curl/_pkg/bin/curl-config $PWD/curl/_pkg/bin/curl-config-BAK

for k in \
    'REQUIRED          COMPONENTS HTTPS'                             \
    '                  COMPONENTS HTTPS'                             \
    'REQUIRED          COMPONENTS HTTPS GOPHER FOO BAR'              \
    '                  COMPONENTS HTTPS GOPHER FOO BAR'              \
    'REQUIRED OPTIONAL_COMPONENTS HTTPS GOPHER FOO BAR'              \
    '         OPTIONAL_COMPONENTS HTTPS GOPHER FOO BAR'              \
    'REQUIRED          COMPONENTS HTTPS OPTIONAL_COMPONENTS FOO BAR' \
    '                  COMPONENTS HTTPS OPTIONAL_COMPONENTS FOO BAR' \
    'REQUIRED          COMPONENTS HTTPS FOO OPTIONAL_COMPONENTS BAR' \
    '                  COMPONENTS HTTPS FOO OPTIONAL_COMPONENTS BAR' \
    'REQUIRED'                                                       \
    ''                                                               \
    'CONFIG'                                                         \
  ; do
  cat > CMakeLists.txt <<EOF
message("")
message("------------------- $k")
cmake_minimum_required(VERSION 3.18)
project(test-dependent LANGUAGES C)
function(curl_dumpvars)
  get_cmake_property(_vars VARIABLES)
  foreach(_var IN ITEMS \${_vars})
    if(_var MATCHES "^CURL")
      message("\${_var} = \${\${_var}}")
    endif()
  endforeach()
endfunction()
find_package(CURL $k)
curl_dumpvars()
EOF
  rm -rf _bld
  cmake -B _bld -DCURL_ROOT=$PWD/curl/_pkg -DCURL_CONFIG_EXECUTABLE=none -DPKG_CONFIG_EXECUTABLE=none -DCMAKE_IGNORE_PREFIX_PATH=/usr/bin
done
```

Reply:
Haha, thanks :) Today is the day to celebrate!

Let me take a closer look maybe in a day or two (some stuff I have to attend to today) but from what I have seen from a quick look, I would still recommend reporting missing optional components. Note that in `CONFIG` mode, there will not be an extra message from the CMake `FindCURL` and the required and optional components found/missing.

But have to say, nifty testing script you have pasted. I tested the "low-tech" way, using an existing project and just fiddling with the `find_package(CURL)` parameters to see what would happen. Unfortunately, Windows CMD is... just bad to script in, so I would probably have to do the test script in Python (this will be more verbose for sure).